File: packages/payment-fee/src/payment_fee/engine.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from pydantic import TypeAdapter, ValidationError

from payment_fee.calculator import FeeCalculator
from payment_fee.data_source import (
    DEFAULT_PAYPAL_URL,
    DEFAULT_STRIPE_URL,
    JsonDataSource,
    _load_documents,
    _location_from_string,
)
from payment_fee.errors import UnknownProvider
from payment_fee.models import (
    BaseQuoteRequest,
    CapabilityInfo,
    MarketInfo,
    PayPalQuoteRequest,
    ProviderInfo,
    QuoteRequest,
    QuoteResponse,
    QuoteSchema,
    StripeQuoteRequest,
)
from payment_fee.providers import PayPalProvider, StripeProvider
from payment_fee.registry import ProviderRegistry
from payment_fee.rules import CompiledFeePlan

_QUOTE_REQUEST_ADAPTER: TypeAdapter[BaseQuoteRequest] = TypeAdapter(QuoteRequest)


class PaymentFeeEngine:
    def __init__(self, registry: ProviderRegistry) -> None:
        self._registry = registry
        self._calculator = FeeCalculator()

    @classmethod
    def from_paths(
        cls,
        paypal: str | None = None,
        stripe: str | None = None,
        validate: bool = False,
    ) -> PaymentFeeEngine:
        registry = ProviderRegistry()
        if paypal:
            registry.register(PayPalProvider.from_paths(paypal, data_ref="local", validate_schema=validate))
        if stripe:
            registry.register(StripeProvider.from_paths(stripe, data_ref="local", validate_schema=validate))
        return cls(registry)

    @classmethod
    def from_remote(
        cls,
        paypal: str | None = None,
        stripe: str | None = None,
        *,
        paypal_data_ref: str | None = None,
        stripe_data_ref: str | None = None,
        cache_dir: str | Path | None = None,
        ttl_seconds: float = 86400.0,
        auto_refresh: bool = True,
        validate: bool = False,
    ) -> PaymentFeeEngine:
        """Build an engine from remote data sources with TTL caching.

        ``paypal`` and ``stripe`` are base URLs. They may contain ``{data_ref}``,
        which is replaced by the corresponding ``*_data_ref`` argument or the
        pinned revision from ``contracts/data-revisions.json`` (falling back to
        ``main``).

        ``cache_dir`` defaults to ``~/.cache/payment-fee`` (or the system temp
        directory). ``ttl_seconds`` defaults to 24 hours.
        """
        if paypal is None and stripe is None:
            paypal = DEFAULT_PAYPAL_URL
            stripe = DEFAULT_STRIPE_URL

        registry = ProviderRegistry()
        if paypal:
            location = _location_from_string("paypal", paypal, paypal_data_ref)
            source = JsonDataSource(
                location,
                cache_dir=Path(cache_dir) if cache_dir else None,
                ttl_seconds=ttl_seconds,
                auto_refresh=auto_refresh,
            )
            docs = _load_documents(source, validate)
            registry.register(
                PayPalProvider.from_documents(
                    core=docs["core"],
                    index=docs["index"],
                    data_ref=location.data_ref,
                    validate_schema=validate,
                )
            )
        if stripe:
            location = _location_from_string("stripe", stripe, stripe_data_ref)
            source = JsonDataSource(
                location,
                cache_dir=Path(cache_dir) if cache_dir else None,
                ttl_seconds=ttl_seconds,
                auto_refresh=auto_refresh,
            )
            docs = _load_documents(source, validate)
            registry.register(
                StripeProvider.from_documents(
                    core=docs["core"],
                    index=docs["index"],
                    data_ref=location.data_ref,
                    validate_schema=validate,
                )
            )
        return cls(registry)

    @classmethod
    def from_documents(
        cls,
        paypal: dict[str, Any] | PayPalProvider | None = None,
        stripe: dict[str, Any] | StripeProvider | None = None,
        validate: bool = False,
    ) -> PaymentFeeEngine:
        registry = ProviderRegistry()
        if paypal:
            if isinstance(paypal, PayPalProvider):
                registry.register(paypal)
            else:
                core = paypal.get("core") or paypal
                index = paypal.get("index")
                schemas = paypal.get("schemas")
                registry.register(
                    PayPalProvider.from_documents(
                        core=core,
                        index=index,
                        schemas=schemas,
                        data_ref="documents",
                        validate_schema=validate,
                    )
                )
        if stripe:
            if isinstance(stripe, StripeProvider):
                registry.register(stripe)
            else:
                core = stripe.get("core") or stripe
                index = stripe.get("index")
                schemas = stripe.get("schemas")
                registry.register(
                    StripeProvider.from_documents(
                        core=core,
                        index=index,
                        schemas=schemas,
                        data_ref="documents",
                        validate_schema=validate,
                    )
                )
        return cls(registry)

    def quote(self, request: BaseQuoteRequest | dict[str, Any]) -> QuoteResponse:
        if isinstance(request, dict):
            provider_id = request.get("provider")
            if provider_id and provider_id not in self._registry.providers():
                raise UnknownProvider(provider_id)
            try:
                request = _QUOTE_REQUEST_ADAPTER.validate_python(request)
            except ValidationError as exc:
                provider_id = request.get("provider") if isinstance(request, dict) else None
                if provider_id:
                    raise UnknownProvider(provider_id) from exc
                raise
        provider = self._registry.get(request.provider)
        plan = provider.compile_rules(request)
        return self._calculator.calculate(request.amount, plan)

    def providers(self) -> list[str]:
        return self._registry.providers()

    def markets(self, provider: str) -> list[MarketInfo]:
        return self._registry.get(provider).markets()

    def capabilities(self, provider: str, market: str) -> CapabilityInfo:
        return self._registry.get(provider).capabilities(market)

    def quote_schema(self, provider: str, market: str) -> QuoteSchema:
        return self._registry.get(provider).quote_schema(market)

    def data_status(self) -> list[ProviderInfo]:
        return self._registry.infos()

    def _compile_rules(
        self,
        provider_id: str,
        request: PayPalQuoteRequest | StripeQuoteRequest,
    ) -> CompiledFeePlan:
        return self._registry.get(provider_id).compile_rules(request)

File: packages/payment-fee-typescript/src/engine.ts
import { calculate } from "./calculator.js";
import type { ExecutableRule } from "./calculator.js";
import {
  DEFAULT_PAYPAL_URL,
  DEFAULT_STRIPE_URL,
  dataLocationFromString,
  JsonDataSource,
} from "./data-source.js";
import { UnknownProvider } from "./errors.js";
import type { QuoteRequest } from "./models.js";
import { PayPalProvider, type PayPalCore } from "./providers/paypal/provider.js";
import { StripeProvider, type StripeCore } from "./providers/stripe/provider.js";

export interface Provider {
  readonly providerId: string;
  compileRules(request: QuoteRequest): ExecutableRule[];
  auditContract(): Record<string, number>;
  assumptions(request: QuoteRequest): string[];
  data(request: QuoteRequest): Record<string, unknown>;
}

function deriveStatus(rules: ExecutableRule[]): string {
  const nonAdditive = rules.filter((r) => (r.behavior ?? "base") !== "additive");
  if (
    nonAdditive.length > 0 &&
    nonAdditive.every((r) => ["free", "included", "waived"].includes(r.behavior ?? "base"))
  ) {
    return "included";
  }
  for (const rule of rules) {
    if (["range", "from", "up_to"].includes(rule.exactness ?? "")) {
      return "range";
    }
  }
  for (const rule of rules) {
    if (
      (rule.exactness ?? "") !== "" &&
      !["exact", "exact_for_public_rate"].includes(rule.exactness ?? "")
    ) {
      return "estimated";
    }
    const status = rule.classification_status ?? "";
    if (
      status !== "" &&
      !["calculable", "calculable_rule", "exact", "exact_for_public_rate"].includes(status)
    ) {
      return "estimated";
    }
  }
  return "exact_for_public_rate";
}

export class PaymentFeeEngine {
  private readonly _providers = new Map<string, Provider>();

  static async fromPaths(args: {
    paypal?: string;
    stripe?: string;
    validate?: boolean;
  }): Promise<PaymentFeeEngine> {
    const { readFile } = await import("node:fs/promises");
    const engine = new PaymentFeeEngine();
    if (args.paypal) {
      const core = JSON.parse(
        await readFile(`${args.paypal}/json/core-fees.json`, "utf-8"),
      ) as PayPalCore;
      engine.register(new PayPalProvider(core));
    }
    if (args.stripe) {
      const core = JSON.parse(
        await readFile(`${args.stripe}/json/core-fees.json`, "utf-8"),
      ) as StripeCore;
      engine.register(new StripeProvider(core));
    }
    return engine;
  }

  static async fromDocuments(args: {
    paypal?: PayPalCore | { core?: PayPalCore };
    stripe?: StripeCore | { core?: StripeCore };
    validate?: boolean;
  }): Promise<PaymentFeeEngine> {
    await Promise.resolve();
    const engine = new PaymentFeeEngine();
    if (args.paypal) {
      const paypal = args.paypal;
      const core = ("core" in paypal ? paypal.core : undefined) ?? (paypal as PayPalCore);
      engine.register(new PayPalProvider(core));
    }
    if (args.stripe) {
      const stripe = args.stripe;
      const core = ("core" in stripe ? stripe.core : undefined) ?? (stripe as StripeCore);
      engine.register(new StripeProvider(core));
    }
    return engine;
  }

  static async fromRemote(args: {
    paypal?: string | null;
    stripe?: string | null;
    paypalDataRef?: string;
    stripeDataRef?: string;
    cacheDir?: string;
    ttlSeconds?: number;
    autoRefresh?: boolean;
    validate?: boolean;
  }): Promise<PaymentFeeEngine> {
    const engine = new PaymentFeeEngine();

    async function load(providerId: "paypal" | "stripe", url: string, dataRef?: string) {
      const location = dataLocationFromString(providerId, url, dataRef);
      const source = new JsonDataSource(location, {
        cacheDir: args.cacheDir,
        ttlSeconds: args.ttlSeconds,
        autoRefresh: args.autoRefresh,
      });
      const core = await source.readJson<PayPalCore | StripeCore>("json/core-fees.json");
      if (providerId === "paypal") {
        engine.register(new PayPalProvider(core as PayPalCore));
      } else {
        engine.register(new StripeProvider(core as StripeCore));
      }
    }

    const paypalUrl =
      args.paypal === undefined && args.stripe === undefined
        ? DEFAULT_PAYPAL_URL
        : (args.paypal ?? null);
    const stripeUrl =
      args.paypal === undefined && args.stripe === undefined
        ? DEFAULT_STRIPE_URL
        : (args.stripe ?? null);
    if (paypalUrl) await load("paypal", paypalUrl, args.paypalDataRef);
    if (stripeUrl) await load("stripe", stripeUrl, args.stripeDataRef);
    return engine;
  }

  register(instance: Provider): void {
    this._providers.set(instance.providerId, instance);
  }

  quote(request: QuoteRequest): Record<string, unknown> {
    this.requireProvider(request.provider);
    const provider = this._providers.get(request.provider)!;
    const rules = provider.compileRules(request);
    const result = calculate(request.amount, request.amount.currency, rules);
    const status = deriveStatus(rules);

    return {
      provider: provider.providerId,
      status,
      amount: result.amount,
      processing_fee: result.processing_fee,
      net_amount: result.net_amount,
      components: result.components,
      matched_rules: result.matched_rules,
      selected_product_id: request.transaction.product_id,
      selected_variant_id: request.transaction.variant_id,
      assumptions: provider.assumptions(request),
      warnings: [],
      data: provider.data(request),
    };
  }

  providers(): string[] {
    return [...this._providers.keys()];
  }

  markets(_provider: string): Record<string, unknown>[] {
    this.requireProvider(_provider);
    return [];
  }

  capabilities(_provider: string, _market: string): Record<string, unknown> {
    this.requireProvider(_provider);
    return {};
  }

  quoteSchema(_provider: string, _market: string): Record<string, unknown> {
    this.requireProvider(_provider);
    return {};
  }

  dataStatus(): Record<string, unknown>[] {
    return [];
  }

  auditContract(): Record<string, number> {
    const result: Record<string, number> = {};
    for (const provider of this._providers.values()) {
      const audit = provider.auditContract();
      for (const [key, value] of Object.entries(audit)) {
        result[key] = (result[key] ?? 0) + value;
      }
    }
    for (const key of [
      "unknown_fields",
      "unknown_condition_dimensions",
      "unknown_condition_operators",
      "unsupported_fee_components",
      "unresolved_schedule_references",
    ]) {
      if (!(key in result)) {
        result[key] = 0;
      }
    }
    return result;
  }

  private requireProvider(provider: string): void {
    if (!this._providers.has(provider)) {
      throw new UnknownProvider(provider);
    }
  }
}
